packages/backend/src/peripherals/database/ForcedTransactionsRepository.ts

```diff
@@ -0,0 +1,409 @@
+import { AssetId, Hash256, json, Timestamp } from '@explorer/types'
+import { Knex } from 'knex'
+import { ForcedTransactionEventRow as EventRow } from 'knex/types/tables'
+import { groupBy, pick } from 'lodash'
+import { MD5 } from 'object-hash'
+
+import { Logger } from '../../tools/Logger'
+
+type EventRecordCandidateBase = {
+  id?: number
+  transactionHash: Hash256
+  timestamp: Timestamp
+}
+
+type WithdrawalMinedEventData = {
+  publicKey: string
+  positionId: bigint
+  amount: bigint
+}
+
+type WithdrawalMinedEventRecordCandidate = EventRecordCandidateBase &
+  WithdrawalMinedEventData & {
+    transactionType: 'withdrawal'
+    eventType: 'mined'
+    blockNumber: number
+  }
+
+type VerifiedEventBase = {
+  eventType: 'verified'
+  stateUpdateId: number
+}
+
+export type WithdrawalVerifiedEventRecordCandidate = EventRecordCandidateBase &
+  VerifiedEventBase & {
+    transactionType: 'withdrawal'
+    blockNumber: number
+  }
+
+type TradeMinedEventData = {
+  publicKeyA: string
+  publicKeyB: string
+  positionIdA: bigint
+  positionIdB: bigint
+  syntheticAssetId: AssetId
+  isABuyingSynthetic: boolean
+  collateralAmount: bigint
+  syntheticAmount: bigint
+  nonce: bigint
+}
+
+type TradeMinedEventRecordCandidate = EventRecordCandidateBase &
+  TradeMinedEventData & {
+    transactionType: 'trade'
+    eventType: 'mined'
+    blockNumber: number
+  }
+
+export type TradeVerifiedEventRecordCandidate = EventRecordCandidateBase &
+  VerifiedEventBase & {
+    transactionType: 'trade'
+    blockNumber: number
+  }
+
+export type EventRecordCandidate =
+  | TradeMinedEventRecordCandidate
+  | TradeVerifiedEventRecordCandidate
+  | WithdrawalMinedEventRecordCandidate
+  | WithdrawalVerifiedEventRecordCandidate
+
+type EventRecord = EventRecordCandidate & {
+  id: EventRow['id']
+}
+
+type EventRowCandidate = Omit<EventRow, 'id'> & {
+  id?: EventRow['id']
+}
+
+function toSerializableJson(data: object): json {
+  return Object.entries(data).reduce((acc, [key, val]) => {
+    return {
+      ...acc,
+      [key]: typeof val === 'bigint' ? val.toString() : val,
+    }
+  }, {})
+}
+
+function hashData(data: object): string {
```

Comment:
This function is a bit useless, because all it does it just calls another function